utils/bitfield/src/lib.rs

```diff
@@ -0,0 +1,453 @@
+            }
+            BitField::Decoded(bv) => {
+                let index = bit as usize;
+                if bv.len() <= index {
+                    bv.resize(index + 1, false);
+                }
+                bv.set(index, true);
+            }
+        }
+    }
+
+    /// Removes bit at bit index provided
+    pub fn unset(&mut self, bit: u64) {
+        match self {
+            BitField::Encoded { set, unset, .. } => {
+                set.remove(&bit);
+                unset.insert(bit);
+            }
+            BitField::Decoded(bv) => {
+                let index = bit as usize;
+                if bv.len() < index {
+                    return;
+                }
+                bv.set(index, false);
+            }
+        }
+    }
+
+    /// Gets the bit at the given index.
+    // TODO this probably should not require mut self and RLE decode bits
+    pub fn get(&mut self, index: u64) -> Result<bool> {
+        match self {
+            BitField::Encoded { set, unset, .. } => {
+                if set.contains(&index) {
+                    return Ok(true);
+                }
+
+                if unset.contains(&index) {
+                    return Ok(false);
+                }
+
+                // Check in encoded for the given bit
+                // This can be changed to not flush changes
+                if let Some(true) = self.as_mut_flushed()?.get(index as usize) {
+                    Ok(true)
+                } else {
+                    Ok(false)
+                }
+            }
+            BitField::Decoded(bv) => {
+                if let Some(true) = bv.get(index as usize) {
+                    Ok(true)
+                } else {
+                    Ok(false)
+                }
+            }
+        }
+    }
+
+    /// Retrieves the index of the first set bit, and error if invalid encoding or no bits set.
+    pub fn first(&mut self) -> Result<u64> {
+        for (i, b) in (0..).zip(self.as_mut_flushed()?.iter()) {
+            if b == &true {
+                return Ok(i);
+            }
+        }
+        // Return error if none found, not ideal but no reason not to match
+        Err("Bitfield has no set bits")
+    }
+
+    fn retrieve_set_indexes<B: FromIterator<u64>>(&mut self, max: usize) -> Result<B> {
+        let flushed = self.as_mut_flushed()?;
+        if flushed.count_ones() > max {
+            return Err("Bits set exceeds max in retrieval");
+        }
+
+        Ok((0..)
+            .zip(self.as_mut_flushed()?.iter())
+            .filter_map(|(i, b)| if b == &true { Some(i) } else { None })
+            .collect())
+    }
+
+    /// Returns a vector of indexes of all set bits
+    pub fn all(&mut self, max: usize) -> Result<Vec<u64>> {
+        self.retrieve_set_indexes(max)
+    }
+
+    /// Returns a Hash set of indexes of all set bits
+    pub fn all_set(&mut self, max: usize) -> Result<FnvHashSet<u64>> {
+        self.retrieve_set_indexes(max)
+    }
```

Comment:
yeah, while I generally would agree, I'm matching the API of the go impl to make our lives easier. Might make sense to expose the iterator of it, but I didn't want to provide more overhead when this is being optimized or used (as it may not even be needed)